lib/rouge/lexers/red.rb

```diff
@@ -0,0 +1,196 @@
+          what what-dir within? word? words-of
+        )
+      end
+
+      def self.actions
+        @actions ||= Set.new %w(
+          absolute add and~ append at back change clear complement copy delete
+          divide even? find form head head? index? insert length? make modify
+          mold move multiply negate next odd? or~ pick poke power put random
+          read reflect remainder remove reverse round select skip sort subtract
+          swap tail tail? take to trim write xor~
+        )
+      end
+
+      def self.natives
+        @natives ||= Set.new %w(
+          all any arccosine arcsine arctangent arctangent2 as as-pair bind
+          browse call catch checksum complement? compose construct context?
+          cosine debase dehex difference do does enbase equal? exclude exp
+          extend func function get get-env greater-or-equal? greater? has in
+          intersect lesser-or-equal? lesser? list-env log-10 log-2 log-e
+          lowercase max min NaN? negative? new-line new-line? not not-equal?
+          now parse positive? prin print reduce remove-each same? set set-env
+          shift sign? sine size? square-root stats strict-equal? tangent throw
+          to-hex to-local-file try type? union unique unset uppercase value?
+          wait zero?
+        )
+      end
+
+      def self.datatypes
+        @datatypes ||= Set.new %w(
+          action! binary! bitset! block! char! datatype! date! email! error!
+          event! file! float! function! get-path! get-word! handle! hash! image!
+          integer! issue! lit-path! lit-word! logic! map! native! none! object!
+          op! pair! paren! path! percent! point! refinement! routine! set-path!
+          set-word! string! tag! time! tuple! typeset! unset! url! vector! word!
+        )
+      end
+
+      def self.operators
+        @operators ||= Set.new %w(
+          + < <> % << = * / - > <= >= ** // == =? >> >>> and or
+        )
+      end
+
+      state :root do
+        # comments
+        rule /;.+$/, Comment::Single
+        rule /comment {.*}/m, Comment::Multiline
+
+        rule /[+-]?[0-9]+\.[0-9]*E[+-]?[0-9]+/, Literal::Number::Float
+        rule /[+-]?\.[0-9]+E[+-]?[0-9]+/, Literal::Number::Float
+        rule /[+-]?[0-9]+E[+-]?[0-9]+/, Literal::Number::Float
+        rule /[+-]?[0-9]*\.[0-9]+?/, Literal::Number::Float
+        rule /[+-]?[0-9]+/, Literal::Number::Integer
+
+        rule /\./, Punctuation
+        rule /,/, Punctuation
+        rule /;/, Punctuation
+        rule /\(/, Punctuation
+        rule /\)/, Punctuation
+        rule /\{/, Punctuation
+        rule /\}/, Punctuation
+        rule /\[/, Punctuation
+        rule /\]/, Punctuation
+        rule /\^\^/, Punctuation
+
+        rule /<[^>]*>/, Name::Label
+        rule /\s+/, Text::Whitespace
+        rule /""".*?"""/m, Literal::String
+        rule /"([^"\\]|\\.)*"/, Literal::String
+        rule /'''.*?'''/m, Literal::String
+        rule /'([^'\\]|\\.)*'/, Literal::String
```

Comment:
May be worthwhile to pull all the string rules out into a state for clarity. `:root` is becoming a large beast.

Reply:
I will need to get back up to speed with Rouge. I don't remember much about states and pushing/popping at this point. It has been a long time since I tried to make this lexer.